### ISSUE TITLE
glfwSwapBuffers seem to be missing a frame

### ISSUE THREAD

**Issue by yukapuka123** [glfw/glfw#674]
Okay so i have the program for both to do the same thing , basically just fill the screen with a background colour 

glfwSwapBuffers() seems to be for some reason missing the first frame 

i have also recreated the same program with GLX and Xlib to demonstrate what is actually supposed to happen 

both of them you can find here, first the glfw program with the absolute basics, you will notice that when it runs it misses the first colour flush i.e blue(which you can see being run in the Xlib and GLX version) :

```c
#include <GL/glew.h> 
#include <GLFW/glfw3.h> 
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

int main () {
   // start GL context and O/S window using the GLFW helper library
   if (!glfwInit ()) {
   fprintf (stderr, "ERROR: could not start GLFW3\n");
    return 1;
 } 

 GLFWwindow* window = glfwCreateWindow (640, 480, "Hello Triangle", NULL, NULL);

 if (!window) {
    fprintf (stderr, "ERROR: could not open window with GLFW3\n");
    glfwTerminate();
    return 1;
 }

 glfwMakeContextCurrent (window);
                 
 // start GLEW extension handler
 glewExperimental = GL_TRUE;
 GLuint err = glewInit();

 if(err!=GLEW_OK){
    fprintf(stderr, "Error: %s\n", glewGetErrorString(err));
    return 1;
 }

 // get version info
 const GLubyte* renderer = glGetString (GL_RENDERER); // get renderer string
 const GLubyte* version = glGetString (GL_VERSION); // version as a string
 printf ("Renderer: %s\n", renderer);
 printf ("OpenGL version supported %s\n", version);

 glClearColor( 0, 0.5, 1, 1 );
 glClear( GL_COLOR_BUFFER_BIT );
 glfwSwapBuffers(window);

 sleep( 1 );

 glClearColor ( 1, 0.5, 0, 1 );
 glClear ( GL_COLOR_BUFFER_BIT );
 glfwSwapBuffers(window);

 sleep( 1 );

 glfwDestroyWindow(window);
 glfwTerminate();
 return 0;
}
```

compile with:

    g++ -o gll (whatever you call it).cpp -lglfw3 -lGL -lGLEW -lm -ldl -lXinerama -lXrandr -lXi -lXcursor -lX11 -lXxf86vm -lpthread

and the GLX Xlib version to show how it is supposed to behave(not mine originally):

```c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <GL/gl.h>
#include <GL/glx.h>

#define GLX_CONTEXT_MAJOR_VERSION_ARB       0x2091
#define GLX_CONTEXT_MINOR_VERSION_ARB       0x2092
typedef GLXContext (*glXCreateContextAttribsARBProc)(Display*, GLXFBConfig, GLXContext,   Bool, const int*);

 // Helper to check for extension string presence.  Adapted from:
 //   http://www.opengl.org/resources/features/OGLextensions/
static bool isExtensionSupported(const char *extList, const char *extension)
{
    const char *start;
    const char *where, *terminator;

    /* Extension names should not have spaces. */
    where = strchr(extension, ' ');
    if (where || *extension == '\0')
    return false;

   /* It takes a bit of care to be fool-proof about parsing the
   OpenGL extensions string. Don't be fooled by sub-strings,
   etc. */
   for (start=extList;;) {
   where = strstr(start, extension);

    if (!where)
      break;

    terminator = where + strlen(extension);

    if ( where == start || *(where - 1) == ' ' )
       if ( *terminator == ' ' || *terminator == '\0' )
       return true;

     start = terminator;
    }

   return false;
}

 static bool ctxErrorOccurred = false;
 static int ctxErrorHandler( Display *dpy, XErrorEvent *ev ) 
 {
   ctxErrorOccurred = true;
   return 0;
 }

 int main(int argc, char* argv[]) 
 {
    Display *display = XOpenDisplay(NULL);

   if (!display)
   {
     printf("Failed to open X display\n");
     exit(1);
   }

  // Get a matching FB config
 static int visual_attribs[] =
  {
      GLX_X_RENDERABLE    , True,
      GLX_DRAWABLE_TYPE   , GLX_WINDOW_BIT,
      GLX_RENDER_TYPE     , GLX_RGBA_BIT,
      GLX_X_VISUAL_TYPE   , GLX_TRUE_COLOR,
      GLX_RED_SIZE        , 8,
      GLX_GREEN_SIZE      , 8,
      GLX_BLUE_SIZE       , 8,
      GLX_ALPHA_SIZE      , 8,
      GLX_DEPTH_SIZE      , 24,
      GLX_STENCIL_SIZE    , 8,
      GLX_DOUBLEBUFFER    , True,
      //GLX_SAMPLE_BUFFERS  , 1,
      //GLX_SAMPLES         , 4,
      None
   };

  int glx_major, glx_minor;

  // FBConfigs were added in GLX version 1.3.
  if ( !glXQueryVersion( display, &glx_major, &glx_minor ) || 
        ( ( glx_major == 1 ) && ( glx_minor < 3 ) ) || ( glx_major < 1 ) )
   {
       printf("Invalid GLX version");
       exit(1);
   }

   printf( "Getting matching framebuffer configs\n" );
   int fbcount;
   GLXFBConfig* fbc = glXChooseFBConfig(display, DefaultScreen(display), visual_attribs, &fbcount);
  if (!fbc)
  {
     printf( "Failed to retrieve a framebuffer config\n" );
     exit(1);
  }
  printf( "Found %d matching FB configs.\n", fbcount );

  // Pick the FB config/visual with the most samples per pixel
  printf( "Getting XVisualInfos\n" );
  int best_fbc = -1, worst_fbc = -1, best_num_samp = -1, worst_num_samp = 999;

  int i;
  for (i=0; i<fbcount; ++i)
  {
     XVisualInfo *vi = glXGetVisualFromFBConfig( display, fbc[i] );
      if ( vi )
     {
        int samp_buf, samples;
        glXGetFBConfigAttrib( display, fbc[i], GLX_SAMPLE_BUFFERS, &samp_buf );
        glXGetFBConfigAttrib( display, fbc[i], GLX_SAMPLES       , &samples  );
  
        printf( "  Matching fbconfig %d, visual ID 0x%2x: SAMPLE_BUFFERS = %d,"
             " SAMPLES = %d\n", 
          i, vi -> visualid, samp_buf, samples );

         if ( best_fbc < 0 || samp_buf && samples > best_num_samp )
         best_fbc = i, best_num_samp = samples;
         if ( worst_fbc < 0 || !samp_buf || samples < worst_num_samp )
        worst_fbc = i, worst_num_samp = samples;
      }
     XFree( vi );
   }

  GLXFBConfig bestFbc = fbc[ best_fbc ];

 // Be sure to free the FBConfig list allocated by glXChooseFBConfig()
 XFree( fbc );

 // Get a visual
 XVisualInfo *vi = glXGetVisualFromFBConfig( display, bestFbc );
 printf( "Chosen visual ID = 0x%x\n", vi->visualid );

 printf( "Creating colormap\n" );
 XSetWindowAttributes swa;
 Colormap cmap;
 swa.colormap = cmap = XCreateColormap( display,
                                     RootWindow( display, vi->screen ), 
                                     vi->visual, AllocNone );
 swa.background_pixmap = None ;
 swa.border_pixel      = 0;
 swa.event_mask        = StructureNotifyMask;

 printf( "Creating window\n" );
 Window win = XCreateWindow( display, RootWindow( display, vi->screen ), 
                          0, 0, 100, 100, 0, vi->depth, InputOutput, 
                          vi->visual, 
                          CWBorderPixel|CWColormap|CWEventMask, &swa );
  if ( !win )
  {
      printf( "Failed to create window.\n" );
      exit(1);
   }

  // Done with the visual info data
  XFree( vi );

  XStoreName( display, win, "GL 3.0 Window" );

  printf( "Mapping window\n" );
  XMapWindow( display, win );

  // Get the default screen's GLX extension list
  const char *glxExts = glXQueryExtensionsString( display,
                                              DefaultScreen( display ) );

 // NOTE: It is not necessary to create or make current to a context before
 // calling glXGetProcAddressARB
    glXCreateContextAttribsARBProc glXCreateContextAttribsARB = 0;
    glXCreateContextAttribsARB = (glXCreateContextAttribsARBProc)
    glXGetProcAddressARB( (const GLubyte *) "glXCreateContextAttribsARB" );

    GLXContext ctx = 0;

  // Install an X error handler so the application won't exit if GL 3.0
  // context allocation fails.
  //
  // Note this error handler is global.  All display connections in all threads
  // of a process use the same error handler, so be sure to guard against other
  // threads issuing X commands while this code is running.
  ctxErrorOccurred = false;
  int (*oldHandler)(Display*, XErrorEvent*) =
     XSetErrorHandler(&ctxErrorHandler);

  // Check for the GLX_ARB_create_context extension string and the function.
  // If either is not present, use GLX 1.3 context creation method.
  if ( !isExtensionSupported( glxExts, "GLX_ARB_create_context" ) ||
      !glXCreateContextAttribsARB )
   {
       printf( "glXCreateContextAttribsARB() not found"
          " ... using old-style GLX context\n" );
       ctx = glXCreateNewContext( display, bestFbc, GLX_RGBA_TYPE, 0, True );
    }

  // If it does, try to get a GL 3.0 context!
  else
  {
      int context_attribs[] =
    {
      GLX_CONTEXT_MAJOR_VERSION_ARB, 3,
      GLX_CONTEXT_MINOR_VERSION_ARB, 0,
     //GLX_CONTEXT_FLAGS_ARB        , GLX_CONTEXT_FORWARD_COMPATIBLE_BIT_ARB,
      None
    };

   printf( "Creating context\n" );
   ctx = glXCreateContextAttribsARB( display, bestFbc, 0,
                                  True, context_attribs );

  // Sync to ensure any errors generated are processed.
   XSync( display, False );
   if ( !ctxErrorOccurred && ctx )
   printf( "Created GL 3.0 context\n" );
   else
   {
    // Couldn't create GL 3.0 context.  Fall back to old-style 2.x context.
    // When a context version below 3.0 is requested, implementations will
    // return the newest context version compatible with OpenGL versions less
    // than version 3.0.
     // GLX_CONTEXT_MAJOR_VERSION_ARB = 1
     context_attribs[1] = 1;
     // GLX_CONTEXT_MINOR_VERSION_ARB = 0
     context_attribs[3] = 0;

    ctxErrorOccurred = false;

    printf( "Failed to create GL 3.0 context"
          " ... using old-style GLX context\n" );
    ctx = glXCreateContextAttribsARB( display, bestFbc, 0, 
                                    True, context_attribs );
       }
    }

 // Sync to ensure any errors generated are processed.
 XSync( display, False );

 // Restore the original error handler
 XSetErrorHandler( oldHandler );

 if ( ctxErrorOccurred || !ctx )
 {
   printf( "Failed to create an OpenGL context\n" );
   exit(1);
 }

 // Verifying that context is a direct context
 if ( ! glXIsDirect ( display, ctx ) )
 {
   printf( "Indirect GLX rendering context obtained\n" );
 }
else
 {
   printf( "Direct GLX rendering context obtained\n" );
 } 

 printf( "Making context current\n" );
 glXMakeCurrent( display, win, ctx );

 glClearColor( 0, 0.5, 1, 1 );
 glClear( GL_COLOR_BUFFER_BIT );
 glXSwapBuffers ( display, win );

 sleep( 1 );

 glClearColor ( 1, 0.5, 0, 1 );
 glClear ( GL_COLOR_BUFFER_BIT );
 glXSwapBuffers ( display, win );

 sleep( 1 );

 glXMakeCurrent( display, 0, 0 );
 glXDestroyContext( display, ctx );

 XDestroyWindow( display, win );
 XFreeColormap( display, cmap );
 XCloseDisplay( display );

   return 0; 
 }
```

compile with 

    g++ -w -o gl3 (whatever you call it).cpp -lGL -lX11

anyway as you will notice that the glfw3 version misses the first frame (i.e the blue one)

the only way to make the blue frame appear in the glf3 version is to add a sleep(1) before this command 

"glClearColor( 0, 0.5, 1, 1 );"

so bottom line is that glfw3SwapBuffer() is missing something there at the beginning and hence missing a frame 

as you will see at the bottom of the GLX X11 version they do exactly the same thing , with identical commands

**Comment by elmindreda** [comment]
Unable to reproduce.

**Comment by elmindreda** [comment]
[Here](https://github.com/glfw/glfw/blob/master/src/glx_context.c#L559) is the relevant implementation of glfwSwapBuffers.

**Comment by yukapuka123** [comment]
really,  unable to reproduce? did you run the two files above?

**Comment by elmindreda** [comment]
Yes.

**Comment by Spartan322** [comment]
This was possibly explained in #671 by tombsar

**Comment by yukapuka123** [comment]
nah i tried it, didnt change anything

**Comment by Spartan322** [comment]
That doesn't mean what he said is wrong

**Comment by elmindreda** [comment]
@yukapuka123 What desktop environment or window manager are you using? What GPU do you have and are you using Mesa or proprietary drivers?

**Comment by yukapuka123** [comment]
that could be the root cause of the problem, but what is why i made the comparison between the x11 glx created window/context and the glfw created windows and context

the x11/glx one runs the same commands i.e 

    glClearColor( 0, 0.5, 1, 1 );
    glClear( GL_COLOR_BUFFER_BIT );
    glXSwapBuffers ( display, win );

    sleep( 1 );

    glClearColor ( 1, 0.5, 0, 1 );
    glClear ( GL_COLOR_BUFFER_BIT );
    glXSwapBuffers ( display, win );

    sleep( 1 );

and doesnt get the same effect

**Comment by yukapuka123** [comment]
i am using gnome cinamon linux mint 
gpu is the titan 780ti 

    OpenGL vendor string: NVIDIA Corporation
    OpenGL renderer string: GeForce GTX 780 Ti/PCIe/SSE2
    OpenGL core profile version string: 4.3.0 NVIDIA 352.63
    OpenGL core profile shading language version string: 4.30 NVIDIA via Cg compiler
    OpenGL core profile context flags: (none)
    OpenGL core profile profile mask: core profile
    OpenGL core profile extensions:
    OpenGL version string: 4.5.0 NVIDIA 352.63
    OpenGL shading language version string: 4.50 NVIDIA
    OpenGL context flags: (none) 
    OpenGL profile mask: (none)
    OpenGL extensions:

**Comment by yukapuka123** [comment]
https://www.youtube.com/watch?v=fcw-MtpTl_U

if it were an issue with my drivers or environment, then it would also happen on the glx/x11 file , but as you can see it doesnt happen

it only happens with glfw

**Comment by elmindreda** [comment]
That video is still private.

**Comment by yukapuka123** [comment]
okay ive set it to unlisted so yo should be able to see it now

**Comment by elmindreda** [comment]
Thank you.

See what happens if you put this before the first clear.

```c
while (!glfwGetWindowAttrib(window, GLFW_VISIBLE))
    ;
```

**Comment by yukapuka123** [comment]
Okay ive tested it three times , recompiled rebuilt etc and it worked each time

 it seems that is the solution to it

definitely felt like something was lagging behind, so it may have been just that

**Comment by elmindreda** [comment]
This is a race between your program and the window manager. GLFW requests a lot more of it for each new window than your minimal X11 program does, so it doesn't have time to finish mapping the window before you swap.

**Comment by yukapuka123** [comment]
i noticed that even putting sleep(1) before the firs glclear also does the job
